File: BillPayments/Features/Payments/Data/PaymentHistoryRepository.swift
import Foundation

protocol PaymentHistoryRepository {
    func restore(userId: String) -> Payments
    func persist(userId: String, value: Payments) async throws
}

private struct PaymentDto: Codable {
    let id: String
    let billerId: String
    let billerName: String
    let categoryId: String
    let account: String
    let amountPaise: Int64
    let paidAtUtc: Double
    let status: String
}

private struct PaymentsDto: Codable {
    let nextId: Int
    let items: [PaymentDto]
}

final class StoredPaymentHistoryRepository: PaymentHistoryRepository {
    private let store: KeyValueStore

    init(store: KeyValueStore) {
        self.store = store
    }

    func restore(userId: String) -> Payments {
        guard let raw = store.read(keyFor(userId)),
              let data = raw.data(using: .utf8),
              let dto = try? JSONDecoder().decode(PaymentsDto.self, from: data)
        else { return .empty }
        return Payments(items: dto.items.map(decode), nextId: dto.nextId)
    }

    func persist(userId: String, value: Payments) async throws {
        let dto = PaymentsDto(
            nextId: value.nextId,
            items: value.items.map {
                PaymentDto(
                    id: $0.id,
                    billerId: $0.billerId,
                    billerName: $0.billerName,
                    categoryId: $0.categoryId,
                    account: $0.account,
                    amountPaise: $0.amountPaise,
                    paidAtUtc: $0.paidAtUtc.timeIntervalSince1970,
                    status: $0.status.rawValue
                )
            }
        )
        let data = try JSONEncoder().encode(dto)
        guard let raw = String(data: data, encoding: .utf8) else { throw StorageError() }
        try await store.write(keyFor(userId), raw)
    }

    private func decode(_ dto: PaymentDto) -> Payment {
        let status = PaymentStatus(rawValue: dto.status) ?? .failed
        return Payment(
            id: dto.id,
            billerId: dto.billerId,
            billerName: dto.billerName,
            categoryId: dto.categoryId,
            account: dto.account,
            amountPaise: dto.amountPaise,
            paidAtUtc: Date(timeIntervalSince1970: dto.paidAtUtc),
            status: status == .processing ? .failed : status
        )
    }

    private func keyFor(_ userId: String) -> String { "payments.\(userId)" }
}

File: app/src/main/java/com/billpayments/features/payments/data/PaymentHistoryRepository.kt
package com.billpayments.features.payments.data

import com.billpayments.core.storage.KeyValueStore
import java.time.Instant
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

interface PaymentHistoryRepository {
    fun restore(userId: String): Payments
    suspend fun persist(userId: String, value: Payments)
}

@Serializable
private data class PaymentDto(
    val id: String,
    val billerId: String,
    val billerName: String,
    val categoryId: String,
    val account: String,
    val amountPaise: Long,
    val paidAtUtc: String,
    val status: String,
)

@Serializable
private data class PaymentsDto(val nextId: Int, val items: List<PaymentDto>)

class StoredPaymentHistoryRepository(private val store: KeyValueStore) : PaymentHistoryRepository {

    override fun restore(userId: String): Payments {
        val raw = store.read(keyFor(userId)) ?: return Payments.EMPTY
        val dto = Json.decodeFromString<PaymentsDto>(raw)
        return Payments(nextId = dto.nextId, items = dto.items.map(::decode))
    }

    override suspend fun persist(userId: String, value: Payments) {
        val dto = PaymentsDto(
            nextId = value.nextId,
            items = value.items.map {
                PaymentDto(
                    it.id, it.billerId, it.billerName, it.categoryId,
                    it.account, it.amountPaise, it.paidAtUtc.toString(), it.status.name,
                )
            },
        )
        store.write(keyFor(userId), Json.encodeToString(dto))
    }

    private fun decode(dto: PaymentDto): Payment {
        val status = PaymentStatus.valueOf(dto.status)
        return Payment(
            id = dto.id,
            billerId = dto.billerId,
            billerName = dto.billerName,
            categoryId = dto.categoryId,
            account = dto.account,
            amountPaise = dto.amountPaise,
            paidAtUtc = Instant.parse(dto.paidAtUtc),
            status = if (status == PaymentStatus.Processing) PaymentStatus.Failed else status,
        )
    }

    private fun keyFor(userId: String) = "payments.$userId"
}
